pkg/controller/apiextensions/composite/reconciler.go

```diff
@@ -0,0 +1,166 @@
+		connection:   NewAPIFilteredSecretPublisher(kube, filterer.GetConnectionSecretKeys()),
+		finalizer:    resource.NewAPIFinalizer(kube, finalizer),
+		log:          log,
+		record:       event.NewAPIRecorder(mgr.GetEventRecorderFor(name)),
+	}
+}
+
+// ComposableReconciler is able to reconcile a member of the composite resource.
+type ComposableReconciler interface {
+	Reconcile(ctx context.Context, cr resource.Composite, composedRef v1.ObjectReference, tmpl v1alpha1.ComposedTemplate) (Observation, error)
+}
+
+// Resolver selects the composition reference with the information given as selector.
+type Resolver interface {
+	ResolveSelector(ctx context.Context, cr resource.Composite) error
+}
+
+// compositeReconciler reconciles the generic CRD that is generated via InfrastructureDefinition.
+type compositeReconciler struct {
+	client       client.Client
+	newComposite func() resource.Composite
+	composed     ComposableReconciler
+	connection   ConnectionPublisher
+	finalizer    resource.Finalizer
+
+	// TODO(muvaf): Implement `Initializer` interface to be satisfied by both
+	// selector resolver and empty connection secret ref defaulter.
+	Resolver
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Reconcile reconciles given custom resource.
+func (r *compositeReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	cr := r.newComposite()
+	if err := r.client.Get(ctx, req.NamespacedName, cr); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), "cannot get composite resource")
+	}
+
+	if meta.WasDeleted(cr) && len(cr.GetResourceReferences()) == 0 {
+		if err := r.finalizer.RemoveFinalizer(ctx, cr); err != nil {
+			return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), "cannot remove finalizer")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.ResolveSelector(ctx, cr); err != nil {
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, "cannot resolve composition selector")
+	}
+	// TODO(muvaf): We should lock the deletion of Composition via finalizer
+	// because its deletion will break the field propagation.
+	comp := &v1alpha1.Composition{}
+	if err := r.client.Get(ctx, meta.NamespacedNameOf(cr.GetCompositionReference()), comp); err != nil {
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, "cannot get the composition")
+	}
+
+	if err := r.finalizer.AddFinalizer(ctx, cr); err != nil {
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, "cannot add finalizer")
+	}
+
+	// We start with empty ObjectRefs and fill them up as they are provisioned.
+	refs := make([]v1.ObjectReference, len(comp.Spec.To))
+	copy(refs, cr.GetResourceReferences())
+	conn := managed.ConnectionDetails{}
+	for i, composedRef := range refs {
+		obs, err := r.composed.Reconcile(ctx, cr, composedRef, comp.Spec.To[i])
```

Comment:
Yeah I had a todo note here in the beginning about this but it would cause `resource changed` errors. In this new structure, it only returns the reference, so it should be doable. Left a note for the next PR.